docs/cip/CIP-8_Matrix_Packages.md

```diff
@@ -0,0 +1,209 @@
+
+TODO exact character restrictions
+
+##### `version`
+```yaml
+version: "2023.09.08-1"
+```
+The version field defines the package's version.
+
+Versioning schemes are a frustratingly subtle and complex topic. The simplest thing we can say about them is that [totally ordered](https://en.wikipedia.org/wiki/Total_order) version schemes are strongly preferable to those without. The format given above is a [CalVer](https://calver.org/) - that is, date-based versioning: `YYYY.MM.DD-X`, where `X` is an incrementing counter.
+
+This format is more of a placeholder to facilitate discussion than a deeply considered proposal, and is at least as as notable for what it isn't as what it is:
+- Dates provide a total order
+- Day-level granularity with an incrementing counter that resets each day is broadly reasonable **IFF** the typical use case for matrix packages will be humans curating periodic releases through centralized processes (incrementing counter entails centralized process). But it may be a poor choice if we expect matrix packages to, for example, be AirFlow outputs.
+- It's not [SemVer](https://semver.org/), because there's no obvious general rule we could use to differentiate between breaking and non-breaking changes to matrix packages.
+- There's no space to wedge in arbitrary build identifiers or hash digests.
+
+TODO flesh it out
+
+##### `license`
+```yaml
+license: "Apache-2.0"
+```
+License contains an [SPDX license identifier](https://spdx.org/licenses/), specifying the license for the package and its contents.
+
+Given all of the open questions ([for example](https://www.infoworld.com/article/3706091/rethinking-open-source-for-ai.html)) about AI training data and ownership, this could end up being a complex area that looks quite unlike the way software licensing has historically worked. However, there does not yet appear to be any alternative clear enough to justify deviating from SPDX standards.
+##### `data`
+`data` contains metadata about the package's dataset - the contents of `matrix.data.parquet`.
+
+This is the meat of the matrix package. It's recommended to read [this appendix](#Appendix%20On%20Applying%20Package%20Management%20to%20AI.md) before going further.
+
+The general principle within the `data` block is to describe correctness properties of the data, without necessarily specifying an exact procedure for arriving at those properties. The (handwavy) goal here is to achieve a distinction analogous to URIs vs. URLs, or types vs. values: enough information to be able to make bounded correctness statements, without being too prescriptive about implementation. It's **where** we're going, rather than **how** to get there.
+
+Exactly how to apply this principle needs consideration for each section.
+
+TODO how much of this can or should be encoded directly in the parquet file? is duplication OK? what would be the reason for having it here instead of or in addition to parquet?
+
+##### `data.embedding`
+```yaml
+  embedding:
```

Comment:
We should also consider DF and generally the underlying hnsw param configuration as changing that will have impacts on reproducing the same results on a different system. Let's apply a similar strategy (nominal vs structural) for the underlying vector search lib.

Reply:
Very glad to see this comment. I've learned just enough about these areas to know they matter, but not enough to really reason about them. Applying some nominal typing strategy to them seemed plausible as well.

I'd love to tease out distinct identifiable properties (c.f. vector dimensionality) present in this particular domain as a way of trying to reduce that cognitive load.

Reply:
See also "[stringly typed](https://wiki.c2.com/?StringlyTyped)" 👿